File: real_data_test/distribution_analysis/src/alignment.rs
use std::collections::HashMap;
use std::hash::RandomState;
use std::rc::Rc;

use crate::domain::DomainContainer;
use crate::reference::Reference;
use crate::domain::Domain;


type RefRc = Rc<Reference>;
type StringHashMap<T> = HashMap<String, T, RandomState>;

#[derive(Debug)]
pub struct Alignment{
    matching_reference: Rc<Reference>,
    bitscore: f64
}

impl DomainContainer for Alignment {
    fn get_domains(&self) -> &Vec<Rc<Domain>> {
        &self.matching_reference.get_domains()
    }
    fn get_name(&self) -> &String {
        &self.matching_reference.get_name()
    }
}

impl Alignment {
    pub fn new(row: &Vec<&str>, ref_hashmap: &StringHashMap<RefRc>) -> Self {
        let matching_reference = ref_hashmap.get(row[1]).unwrap().clone();
        let bitscore = row[11].parse().unwrap();
        Self { matching_reference, bitscore}
    }
    pub fn get_bitscore(&self) -> &f64 {
        &self.bitscore
    }
}

File: real_data_test/distribution_analysis/differential_distribution_classes/alignment.py
from differential_distribution_classes.domain import DomainContainer, Domain
from differential_distribution_classes.reference import Reference

class Alignment(DomainContainer):
    def __init__(self, row: list[str], ref_dict: dict[str, Reference]):
        self.matching_reference = ref_dict[row[1]]
        self.bitscore = float(row[11])
        self.start = int(row[8])
        self.end = int(row[9])
        self.al_len = int(row[3])

    def get_bitscore(self) -> float:
        return self.bitscore
    
    def get_coverage(self) -> float:
        return (float(self.al_len) / float(self.matching_reference.get_length()))
    
    def get_domains(self) -> list[Domain]:
        aligned_domains = list()
        for domain in self.matching_reference.get_domains():
            if (domain.start < self.end) and (domain.end > self.start):
                aligned_domains.append(domain)            
        return aligned_domains
    
    def get_name(self) -> str:
        return self.matching_reference.get_name()
    
    def get_cluster(self) -> int:
        return self.matching_reference.get_cluster()